### PR TITLE
Create party member connections

### PR DESCRIPTION
**Pull request by ruancomelli**
## Summary by CodeRabbit

- **New Features**
	- Introduced a function to dynamically list and style party members with detailed information and optional summaries.
- **Improvements**
	- Enhanced party and player character creation forms with additional elements and error handling.
	- Improved logging, data handling, and function signatures for several capabilities.
	- Updated serialization and database connection settings for better performance.
- **Documentation**
	- Expanded README with detailed architecture explanations and error handling practices.

### CODE REVIEW

**Comment by ruancomelli** [comment]
This is an experiment review for experiment `multiple_mermaid_diagram_types`
Run ID: `multiple_mermaid_diagram_types/run_2024-09-30T09-03-13_v1-23-0-35-g6ed6a52fb`
The benchmark review for this pull request can be found at https://github.com/sourcery-ai-experiments/party-connections/pull/7.
This pull request was cloned from `https://github.com/dkittle/party-connections/pull/6`. _(Note: the URL is not a link to avoid triggering a notification on the original pull request.)_

<details>
<summary>Experiment configuration</summary>

```yaml
review_config:
  # User configuration for the review
  # - benchmark - use the user config from the benchmark reviews
  # - <value> - use the value directly
  user_review_config:
    enable_ai_review: true
    enable_rule_comments: false

    enable_complexity_comments: false
    enable_security_comments: false
    enable_tests_comments: false
    enable_comment_suggestions: false
    enable_functionality_review: false

    enable_pull_request_summary: false
    enable_review_guide: true

    enable_approvals: false

  ai_review_config:
    # The model responses to use for the experiment
    # - benchmark - use the model responses from the benchmark reviews
    # - llm - call the language model to generate responses
    model_responses:
      comments_model: benchmark
      comment_area_model: benchmark
      comment_validation_model: benchmark
      comment_suggestion_model: benchmark
      complexity_model: benchmark
      functionality_model: benchmark
      security_model: benchmark
      tests_model: benchmark
      pull_request_summary_model: benchmark
      review_guide_model: llm
      overall_comments_model: benchmark

# The pull request dataset to run the experiment on
pull_request_dataset:
  # Sourcery core examples:
- https://github.com/sourcery-ai/core/pull/4607
- https://github.com/sourcery-ai/core/pull/4631
- https://github.com/sourcery-ai/core/pull/4647
  # CodeRabbit examples:
- https://github.com/2lambda123/-Orange-OpenSource-oorobot/pull/15
- https://github.com/2lambda123/galaxyproject-galaxy/pull/12
- https://github.com/a0v0/avtoolz/pull/79
- https://github.com/adityask98/Hotaru/pull/10
- https://github.com/agdas/vscode/pull/2
- https://github.com/agluszak/hirschgarten/pull/2
- https://github.com/alexsnow348/insightface/pull/46
- https://github.com/alikuxac/utilities/pull/10
- https://github.com/AlphaDev87/timba-api/pull/49
- https://github.com/AngeloTadeucci/Maple2/pull/239
- https://github.com/AngeloTadeucci/Maple2.File/pull/36
- https://github.com/AngeloTadeucci/Maple2/pull/233
  # Examples where CodeRabbit does not generate diagrams
- https://github.com/baptisteArno/typebot.io/pull/1778
- https://github.com/btipling/foundations/pull/33
- https://github.com/btipling/foundations/pull/31
- https://github.com/chintu-777/jaeger/pull/1
- https://github.com/coji/remix-docs-ja/pull/55
- https://github.com/DaveMBush/SmartNgRX/pull/622
- https://github.com/DaveMBush/SmartNgRX/pull/481
- https://github.com/dkittle/party-connections/pull/6
- https://github.com/Drajad-Kusuma-Adi/onstudy-backend/pull/6
- https://github.com/imaami/libcanth/pull/2
  # Requested by Tim
- https://github.com/2lambda123/-Orange-OpenSource-oorobot/pull/15
- https://github.com/2lambda123/galaxyproject-galaxy/pull/12
- https://github.com/a0v0/avtoolz/pull/79
- https://github.com/adityask98/Hotaru/pull/10
- https://github.com/agdas/vscode/pull/2
- https://github.com/agluszak/hirschgarten/pull/2
- https://github.com/4DNucleome/PartSeg/pull/1183
- https://github.com/ably/ably-cocoa/pull/1973
- https://github.com/ably-labs/ably-chat-swift/pull/54
- https://github.com/ably-labs/ably-chat-swift/pull/35
  # A ton of new classes:
- https://github.com/sourcery-ai/core/pull/4618
  # Including database changes to evaluate entity-relationship diagrams
- https://github.com/sourcery-ai/core/pull/4579
  # TODO: find some with Alembic
  # Brilliant existing examples:
- https://github.com/sourcery-ai/core/pull/4680
- https://github.com/sourcery-ai/core/pull/4684

# Questions to ask to label the review comments
review_comment_labels: []
# - label: correct
#   question: Is this comment correct?

# Benchmark reviews generated by running
#   python -m scripts.experiment benchmark <experiment_name>
benchmark_reviews:
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4607
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/374
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4631
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/375
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4647
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/376
- dataset_pull_request: https://github.com/2lambda123/-Orange-OpenSource-oorobot/pull/15
  review_pull_request: https://github.com/sourcery-ai-experiments/-Orange-OpenSource-oorobot/pull/5
- dataset_pull_request: https://github.com/2lambda123/galaxyproject-galaxy/pull/12
  review_pull_request: https://github.com/sourcery-ai-experiments/galaxyproject-galaxy/pull/13
- dataset_pull_request: https://github.com/adityask98/Hotaru/pull/10
  review_pull_request: https://github.com/sourcery-ai-experiments/Hotaru/pull/14
- dataset_pull_request: https://github.com/agdas/vscode/pull/2
  review_pull_request: https://github.com/sourcery-ai-experiments/vscode/pull/14
- dataset_pull_request: https://github.com/agluszak/hirschgarten/pull/2
  review_pull_request: https://github.com/sourcery-ai-experiments/hirschgarten/pull/13
- dataset_pull_request: https://github.com/alikuxac/utilities/pull/10
  review_pull_request: https://github.com/sourcery-ai-experiments/utilities/pull/4
- dataset_pull_request: https://github.com/AlphaDev87/timba-api/pull/49
  review_pull_request: https://github.com/sourcery-ai-experiments/timba-api/pull/12
- dataset_pull_request: https://github.com/AngeloTadeucci/Maple2/pull/239
  review_pull_request: https://github.com/sourcery-ai-experiments/Maple2/pull/23
- dataset_pull_request: https://github.com/AngeloTadeucci/Maple2.File/pull/36
  review_pull_request: https://github.com/sourcery-ai-experiments/Maple2.File/pull/12
- dataset_pull_request: https://github.com/AngeloTadeucci/Maple2/pull/233
  review_pull_request: https://github.com/sourcery-ai-experiments/Maple2/pull/24
- dataset_pull_request: https://github.com/baptisteArno/typebot.io/pull/1778
  review_pull_request: https://github.com/sourcery-ai-experiments/typebot.io/pull/7
- dataset_pull_request: https://github.com/btipling/foundations/pull/33
  review_pull_request: https://github.com/sourcery-ai-experiments/foundations/pull/13
- dataset_pull_request: https://github.com/btipling/foundations/pull/31
  review_pull_request: https://github.com/sourcery-ai-experiments/foundations/pull/14
- dataset_pull_request: https://github.com/chintu-777/jaeger/pull/1
  review_pull_request: https://github.com/sourcery-ai-experiments/jaeger/pull/7
- dataset_pull_request: https://github.com/coji/remix-docs-ja/pull/55
  review_pull_request: https://github.com/sourcery-ai-experiments/remix-docs-ja/pull/7
- dataset_pull_request: https://github.com/DaveMBush/SmartNgRX/pull/622
  review_pull_request: https://github.com/sourcery-ai-experiments/SmartNgRX/pull/13
- dataset_pull_request: https://github.com/DaveMBush/SmartNgRX/pull/481
  review_pull_request: https://github.com/sourcery-ai-experiments/SmartNgRX/pull/14
- dataset_pull_request: https://github.com/dkittle/party-connections/pull/6
  review_pull_request: https://github.com/sourcery-ai-experiments/party-connections/pull/7
- dataset_pull_request: https://github.com/Drajad-Kusuma-Adi/onstudy-backend/pull/6
  review_pull_request: https://github.com/sourcery-ai-experiments/onstudy-backend/pull/7
- dataset_pull_request: https://github.com/imaami/libcanth/pull/2
  review_pull_request: https://github.com/sourcery-ai-experiments/libcanth/pull/7
- dataset_pull_request: https://github.com/2lambda123/-Orange-OpenSource-oorobot/pull/15
  review_pull_request: https://github.com/sourcery-ai-experiments/-Orange-OpenSource-oorobot/pull/6
- dataset_pull_request: https://github.com/2lambda123/galaxyproject-galaxy/pull/12
  review_pull_request: https://github.com/sourcery-ai-experiments/galaxyproject-galaxy/pull/14
- dataset_pull_request: https://github.com/adityask98/Hotaru/pull/10
  review_pull_request: https://github.com/sourcery-ai-experiments/Hotaru/pull/15
- dataset_pull_request: https://github.com/agdas/vscode/pull/2
  review_pull_request: https://github.com/sourcery-ai-experiments/vscode/pull/15
- dataset_pull_request: https://github.com/agluszak/hirschgarten/pull/2
  review_pull_request: https://github.com/sourcery-ai-experiments/hirschgarten/pull/14
- dataset_pull_request: https://github.com/4DNucleome/PartSeg/pull/1183
  review_pull_request: https://github.com/sourcery-ai-experiments/PartSeg/pull/7
- dataset_pull_request: https://github.com/ably/ably-cocoa/pull/1973
  review_pull_request: https://github.com/sourcery-ai-experiments/ably-cocoa/pull/6
- dataset_pull_request: https://github.com/ably-labs/ably-chat-swift/pull/54
  review_pull_request: https://github.com/sourcery-ai-experiments/ably-chat-swift/pull/11
- dataset_pull_request: https://github.com/ably-labs/ably-chat-swift/pull/35
  review_pull_request: https://github.com/sourcery-ai-experiments/ably-chat-swift/pull/12
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4579
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/377

```

</details>

**Comment by SourceryAI** [comment]
## Reviewer's Guide by Sourcery

This pull request implements significant changes to the Party Connections application, focusing on improving data models, enhancing error handling, and refining the user interface. The changes include the introduction of value classes for better type safety, updates to database operations, and improvements to the party and character creation processes.

### Diagrams

#### Class diagram for new value classes

```mermaid
classDiagram
    class PlayerCharacter {
        +Name name
        +Ancestry ancestry
        +BaseClass baseClass
        +PartyRole role
        +Background background
        +Backstory backstory
        +PartyId partyId
        +PlayerCharacterId id
    }
    class Name {
        +String value
    }
    class Ancestry {
        +String value
    }
    class BaseClass {
        +String value
    }
    class PartyRole {
        +String value
    }
    class Background {
        +String value
    }
    class Backstory {
        +String value
    }
    class PartyId {
        +String value
    }
    class PlayerCharacterId {
        +String value
    }
    PlayerCharacter *-- Name
    PlayerCharacter *-- Ancestry
    PlayerCharacter *-- BaseClass
    PlayerCharacter *-- PartyRole
    PlayerCharacter *-- Background
    PlayerCharacter *-- Backstory
    PlayerCharacter *-- PartyId
    PlayerCharacter *-- PlayerCharacterId
```

#### Sequence diagram for improved error handling in routes

```mermaid
sequenceDiagram
    participant C as Client
    participant R as Route Handler
    participant DB as Database
    participant ER as Error Response
    C->>R: Request
    R->>DB: Query
    alt Successful query
        DB->>R: Data
        R->>C: HTML Response
    else Error occurs
        DB->>R: Error
        R->>ER: Create Error Message
        ER->>R: Formatted Error
        R->>C: Error Response (400 or 404)
```

### File-Level Changes

| Change | Details | Files |
| ------ | ------- | ----- |
| Introduced value classes for improved type safety and validation | <ul><li>Created value classes for Name, Ancestry, BaseClass, PartyRole, Background, Backstory, and PlayerCharacterId</li><li>Implemented validation logic in value classes (e.g., non-empty string checks)</li><li>Updated existing models and functions to use the new value classes</li></ul> | `src/main/kotlin/ca/kittle/capabilities/pc/models/PlayerCharacter.kt`<br/>`src/main/kotlin/ca/kittle/capabilities/party/models/Party.kt`<br/>`src/main/kotlin/ca/kittle/capabilities/party/models/PartyMember.kt` |
| Enhanced error handling and response management | <ul><li>Implemented try-catch blocks in route handlers to catch and respond to exceptions</li><li>Added StatusErrorMessage object for consistent error message formatting</li><li>Updated MongoDB connection settings with timeout configurations</li></ul> | `src/main/kotlin/ca/kittle/plugins/Routing.kt`<br/>`src/main/kotlin/ca/kittle/plugins/Databases.kt`<br/>`src/main/kotlin/ca/kittle/capabilities/party/CreateParty.kt`<br/>`src/main/kotlin/ca/kittle/capabilities/party/ShowParty.kt` |
| Improved party and character management functionality | <ul><li>Created new ShowPartyMembers route and component for displaying party members</li><li>Updated CreatePartyForm with error handling and required fields</li><li>Implemented PartyMember model and related database operations</li></ul> | `src/main/kotlin/ca/kittle/capabilities/party/ShowPartyMembers.kt`<br/>`src/main/kotlin/ca/kittle/capabilities/party/components/CreatePartyForm.kt`<br/>`src/main/kotlin/ca/kittle/capabilities/party/components/ListPartyMembers.kt`<br/>`src/main/kotlin/ca/kittle/db/models/PartyMemberEntity.kt` |
| Updated serialization and database operations | <ul><li>Configured JSON serialization with custom settings</li><li>Implemented new database query for retrieving party members with backstory summaries</li><li>Updated existing database operations to work with new value classes</li></ul> | `src/main/kotlin/ca/kittle/plugins/Serialization.kt`<br/>`src/main/kotlin/ca/kittle/capabilities/party/ShowParty.kt`<br/>`src/main/kotlin/ca/kittle/db/models/PartyMemberEntity.kt` |
| Enhanced user interface and client-side functionality | <ul><li>Added response-targets.js for improved error handling in the UI</li><li>Updated HTML templates to work with new models and error handling</li><li>Implemented auto-refresh functionality for party member list</li></ul> | `src/main/resources/static/response-targets.js`<br/>`src/main/kotlin/ca/kittle/web/Index.kt`<br/>`src/main/kotlin/ca/kittle/capabilities/party/components/ListPartyMembers.kt` |

---

<details>
<summary>Tips and commands</summary>

#### Interacting with Sourcery

- **Trigger a new review:** Comment `@sourcery-ai review` on the pull request.
- **Continue discussions:** Reply directly to Sourcery's review comments.
- **Generate a GitHub issue from a review comment:** Ask Sourcery to create an
  issue from a review comment by replying to it.

#### Customizing Your Experience

Access your [dashboard](https://app.sourcery.ai) to:
- Enable or disable review features such as the Sourcery-generated pull request
  summary, the reviewer's guide, and others.
- Change the review language.
- Add, remove or edit custom review instructions.
- Adjust other review settings.

#### Getting Help

- [Contact our support team](mailto:support@sourcery.ai) for questions or feedback.
- Visit our [documentation](https://docs.sourcery.ai) for detailed guides and information.
- Keep in touch with the Sourcery team by following us on [X/Twitter](https://x.com/SourceryAI), [LinkedIn](https://www.linkedin.com/company/sourcery-ai/) or [GitHub](https://github.com/sourcery-ai).

</details>

File: src/main/kotlin/ca/kittle/plugins/Databases.kt [hunk 1]
```diff
@@ -39,9 +42,20 @@ fun Application.connectToMongoDB(): MongoDatabase {
     val databaseName = environment.config.tryGetString("db.mongo.database.name") ?: "PartyConnections"
 
     val credentials = user.let { userVal -> password?.let { passwordVal -> "$userVal:$passwordVal@" } }.orEmpty()
-    val uri = "mongodb://$credentials$host:$port/?replicaSet=rs0&maxPoolSize=$maxPoolSize&w=majority"
+    val uri = "mongodb://$credentials$host:$port/?replicaSet=rs0&maxPoolSize=$maxPoolSize&w=majority&connectTimeoutMS=1500"
 
-    val mongoClient = MongoClients.create(uri)
+//    val mongoClient = MongoClients.create(uri)
+    val mongoClient = MongoClients.create(
```
Comment (SourceryAI) [inline src/main/kotlin/ca/kittle/plugins/Databases.kt:48]: **suggestion (performance):** Consider adjusting connection timeout values

The current timeout values (1 second for connection, 15 seconds for read, 1 second for server selection) might be too short for some scenarios. Consider making these configurable or adjusting based on expected network conditions and database response times.

File: src/main/resources/static/response-targets.js [hunk 2]
```diff
@@ -0,0 +1,129 @@
+(function() {
```
Comment (SourceryAI) [inline src/main/resources/static/response-targets.js:1]: **suggestion:** Consider simplifying frontend error handling

The response-targets.js file adds significant complexity to frontend error handling. While it provides fine-grained control, consider if this level of complexity is necessary. If it is, ensure it's well-documented for future maintenance.

File: src/main/kotlin/ca/kittle/capabilities/party/ShowPartyMembers.kt [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+            }
+            partyMembers?.apply {
+                call.respondHtml {
+                    body {
+                        listPartyMembers(PartyId(id), partyMembers)
+                    }
+                }
+            } ?: call.respond(HttpStatusCode.NotFound, StatusErrorMessage("No party found with the given id"))
+        } catch (e: Exception) {
+            call.respond(HttpStatusCode.BadRequest, StatusErrorMessage("Failed to retrieve party and party members"))
```
Comment (SourceryAI) [inline src/main/kotlin/ca/kittle/capabilities/party/ShowPartyMembers.kt:32]: **suggestion:** Consider more specific error messages

The error message 'Failed to retrieve party and party members' is quite generic. Consider providing more specific error messages that indicate the nature of the failure (e.g., database connection issue, invalid ID format, etc.) to aid in debugging and user understanding.

File: src/main/kotlin/ca/kittle/capabilities/party/models/Party.kt [hunk 4]
```diff
@@ -1,13 +1,22 @@
...
-    val name: String,
-    @Contextual @SerialName("_id")
-    val id: String = UUID.randomUUID().toString(),
+    val name: Name,
+    val partyMembers: List<PartyMember> = listOf(),
+    @SerialName("_id")
+    val id: PartyId = PartyId(),
 )
+
+object NewId {
```
Comment (SourceryAI) [inline src/main/kotlin/ca/kittle/capabilities/party/models/Party.kt:16]: **suggestion:** Consider using a more descriptive name for NewId

The name 'NewId' is somewhat vague. Consider a more descriptive name like 'GenerateUniqueId' or 'CreateRandomUUID' to better convey its purpose.

**Review by SourceryAI — COMMENTED** [review]
Hey @ruancomelli - I've reviewed your changes - here's some feedback:

**Overall Comments**:
- Consider adding more comprehensive unit tests, especially for the new value classes and refactored database interactions.
- Some introduced concepts (like the `Validated` sealed class) appear incomplete or unused. Consider either fully implementing these or removing them if not immediately necessary.

<details>
<summary>Here's what I looked at during the review</summary>

- 🟡 **General issues**: 4 issues found
- 🟢 **Documentation**: all looks good
</details>

***

[LangSmith trace](https://smith.langchain.com/o/7d8518ad-6dd3-5c7b-96d2-757769296335/projects/p/88560225-5072-48bc-967c-dc54da53f265/r/f04ddaec-0cd0-4c03-8848-0c93e1ec452b?poll=true)

<sub>
Help me be more useful! Please click 👍 or 👎 on each comment and I'll use the feedback to improve your reviews.
</sub>

**Pull request closed without merging**